agent/eni/iphelperwrapper/iphelper_windows.go

```diff
@@ -0,0 +1,155 @@
+	}
+
+	// Initial notification is sent immediately once the callback is registered.
+	// If the same is not received within a defined timeout then we assume an error and close the monitor
+	for {
+		select {
+		case event := <-notificationChannel:
+			if event == InitialNotification {
+				log.Info("IPHelper : NotifyIPInterfaceChange Windows API call successful. ENI Monitor has been started.")
+				return nil
+			}
+		case <-time.After(InitialNotificationTimeout):
+			log.Errorf("IPHelper : Initial notification not received within timeout. Closing the Monitor.")
+			if retErr := monitor.Close(); retErr != nil {
+				return errors.Wrapf(retErr, "initial notification not received as well")
+			}
+			return errors.Errorf("initial notification not received while starting monitor")
+		}
+	}
+}
+
+// This is the callback function invoked by NotifyIPInterfaceChange Windows API
+// The api returns the information about changed interface as a MIB_IPInterface_Row struct
+// The type of event can be inferred using the notificationType. Since we are concerned about addition of new ENIs,
+// therefore we will pass information to notification channel when the event is of interface addition i.e.1
+func callback(callerContext, row, notificationType uintptr) uintptr {
+
+	// Notification type 1 means addition of a device. InterfaceIndex would always be a positive value
+	if notificationType == 1 {
+		rowdata := *(*MibIPInterfaceRow)(unsafe.Pointer(row))
+		log.Debugf("Interface added with index : %v", rowdata.InterfaceIndex)
+		notificationChannel <- int(rowdata.InterfaceIndex)
+	} else if notificationType == 3 {
+		// Notification type = 3 means Initial notification. This block would be executed only once i.e. when agent starts
+		// Initial notification is sent immediately after registration.
+		// Only after sending this, any other types of notification can be sent
+		// Sending this notification on separate goroutine inorder to avoid deadlock with calling function
+		go func() {
+			// In order to confirm the registration of callback, we will send a negative number on the notification channel
+			notificationChannel <- InitialNotification
+		}()
+	}
+
+	return 0
+}
+
+// This method can be called to cancel the notification callback.
+// Alternatively the notification callback would be cancelled when application exits
+func (monitor *eniMonitor) Close() (err error) {
+	log.Info("IPHelper : Cancelling the interface change notification callback")
+	if err := monitor.cancelNotifications(); err != nil {
+		return err
+	}
+	return nil
+}
+
+// This method is used to cancel the notification callback for any changes in the interfaces attached to the system
+// If the call is successful then res is 0. Otherwise we will return the corresponding error.
+func (monitor *eniMonitor) cancelNotifications() error {
+	res, _, _ := funcCancelMibChangeNotify2(uintptr(monitor.notificationHandle))
```

Comment:
Error would not yield any useful information in this case.
"retVal" would contain the error number in case there is an error otherwise it would be 0.